File: egg-niuke-training/08-is-balanced-tree.py
"""
NC62 判断是不是平衡二叉树
https://www.nowcoder.com/practice/8b3b95850edb4115918ecebdf1b4d222

输入一棵节点数为 n 二叉树，判断该二叉树是否是平衡二叉树。
平衡二叉树：任何节点的两个子树的高度差不超过1。

时间复杂度：O(n)
空间复杂度：O(h)
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_balanced(root):
    """
    判断是否为平衡二叉树
    """
    def height(node):
        if not node:
            return 0
        
        left_height = height(node.left)
        if left_height == -1:
            return -1
        
        right_height = height(node.right)
        if right_height == -1:
            return -1
        
        # 检查平衡性
        if abs(left_height - right_height) > 1:
            return -1
        
        return max(left_height, right_height) + 1
    
    return height(root) != -1

# 测试
if __name__ == "__main__":
    # 平衡二叉树
    root1 = TreeNode(1)
    root1.left = TreeNode(2)
    root1.right = TreeNode(3)
    root1.left.left = TreeNode(4)
    root1.left.right = TreeNode(5)
    
    print(f"树1是否平衡: {is_balanced(root1)}")
    
    # 不平衡二叉树
    root2 = TreeNode(1)
    root2.left = TreeNode(2)
    root2.left.left = TreeNode(3)
    root2.left.left.left = TreeNode(4)
    
    print(f"树2是否平衡: {is_balanced(root2)}")

File: egg-niuke-training/08-is-balanced-tree.cpp
/**
 * NC62 判断是不是平衡二叉树
 * https://www.nowcoder.com/practice/8b3b95850edb4115918ecebdf1b4d222
 * 
 * 输入一棵节点数为 n 二叉树，判断该二叉树是否是平衡二叉树。
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(h)
 */

#include <iostream>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool IsBalanced(TreeNode* root) {
        return height(root) != -1;
    }

private:
    int height(TreeNode* node) {
        if (!node) return 0;
        
        int leftHeight = height(node->left);
        if (leftHeight == -1) return -1;
        
        int rightHeight = height(node->right);
        if (rightHeight == -1) return -1;
        
        // 检查平衡性
        if (abs(leftHeight - rightHeight) > 1) {
            return -1;
        }
        
        return max(leftHeight, rightHeight) + 1;
    }
};

int main() {
    Solution solution;
    
    // 平衡二叉树
    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    root1->left->left = new TreeNode(4);
    root1->left->right = new TreeNode(5);
    
    cout << "树1是否平衡: " << (solution.IsBalanced(root1) ? "是" : "否") << endl;
    
    // 不平衡二叉树
    TreeNode* root2 = new TreeNode(1);
    root2->left = new TreeNode(2);
    root2->left->left = new TreeNode(3);
    root2->left->left->left = new TreeNode(4);
    
    cout << "树2是否平衡: " << (solution.IsBalanced(root2) ? "是" : "否") << endl;
    
    return 0;
}

